### ISSUE TITLE
Use Kokkos Sparse Matrix for Constraint Assembly

### ISSUE THREAD

**Issue by ddement** [Exawind/openturbine#181]
Task:  Use KokkosSparse for large matrix operations in the constraint assembly procedure.  

Why:  This will improve the performance of the assembly procedure for medium to large problems by using sparse matrix multiplication.  It is also one step along the path to using sparse matrices for the total system solve.  

Done means:  The AssembleConstraints function uses sparse matrices internally and copies the result into the existing dense system matrix.

**Comment by ddement** [comment]
This has been implemented in #182 .  Here is the new profiing report:
```
BEGIN KOKKOS PROFILING REPORT:
TOTAL TIME: 31.3758 seconds
TOP-DOWN TIME TREE:
<average time> <percent of total time> <percent time in Kokkos> <percent MPI imbalance> <remainder> <kernels per second> <number of calls> <name> [type]
===================
|-> 2.99e+01 sec 95.3% 7.1% 0.0% 0.3% 9.74e+01 11 Step [region]
|   |-> 2.75e+01 sec 87.8% 1.9% 0.0% 0.0% 5.59e+00 22 Solve System [region]
|   |   |-> 2.71e+01 sec 86.3% 0.3% 0.0% 99.7% 1.62e+00 22 KokkosLapack::gesv[TPL_CUSOLVER,double] [region]
|   |   |   |-> 7.34e-02 sec 0.2% 100.0% 0.0% ------ 22 Kokkos::View::initialization [getrf workspace] via memset [for]
|   |   |-> 4.00e-01 sec 1.3% 100.0% 0.0% ------ 22 PreconditionSt [for]
|   |   |-> 3.67e-02 sec 0.1% 100.0% 0.0% ------ 22 PostconditionSt [for]
|   |-> 1.54e+00 sec 4.9% 80.6% 0.0% 19.3% 7.29e+02 22 Assemble System [region]
|   |   |-> 3.76e-01 sec 1.2% 100.0% 0.0% ------ 22 Kokkos::ViewFill-2D [for]
|   |   |-> 2.37e-01 sec 0.8% 100.0% 0.0% ------ 22 PopulateSparseIndices [for]
|   |   |-> 1.17e-01 sec 0.4% 99.8% 0.0% 0.2% 1.89e+02 22 Assemble Elastic Stiffness Matrix [region]
|   |   |   |-> 1.16e-01 sec 0.4% 100.0% 0.0% ------ 22 IntegrateElasticStiffnessMatrix [for]
|   |   |-> 1.04e-01 sec 0.3% 99.8% 0.0% 0.2% 2.11e+02 22 Assemble Inertial Stiffness Matrix [region]
|   |   |   |-> 1.04e-01 sec 0.3% 100.0% 0.0% ------ 22 IntegrateMatrix [for]
|   |   |-> 1.04e-01 sec 0.3% 99.8% 0.0% 0.2% 2.12e+02 22 Assemble Mass Matrix [region]
|   |   |   |-> 1.04e-01 sec 0.3% 100.0% 0.0% ------ 22 IntegrateMatrix [for]
|   |   |-> 1.03e-01 sec 0.3% 99.8% 0.0% 0.2% 2.13e+02 22 Assemble Gyroscopic Inertia Matrix [region]
|   |   |   |-> 1.03e-01 sec 0.3% 100.0% 0.0% ------ 22 IntegrateMatrix [for]
|   |   |-> 4.46e-02 sec 0.1% 100.0% 0.0% ------ 22 KokkosSparse::SpAdd:Numeric::InputNotSorted [for]
|   |-> 7.34e-01 sec 2.3% 46.9% 0.0% 6.2% 1.32e+03 22 Assemble Constraints [region]
|   |   |-> 3.45e-01 sec 1.1% 0.0% 0.0% 100.0% 0.00e+00 22 KokkosSparse::spmv[TPL_CUSPARSE,double] [region]
|   |   |-> 2.38e-01 sec 0.8% 100.0% 0.0% ------ 22 PopulateSparseIndices [for]
|   |   |-> 3.39e-02 sec 0.1% 100.0% 0.0% ------ 22 UpdateIterationMatrix [for]
|   |   |-> 3.37e-02 sec 0.1% 100.0% 0.0% ------ 22 Kokkos::ViewFill-2D [for]
|-> 4.81e-02 sec 0.2% 100.0% 0.0% ------ 600 Kokkos::ViewFill-3D [for]

KOKKOS HOST SPACE:
===================
MAX MEMORY ALLOCATED: 12035.2 kB
ALLOCATIONS AT TIME OF HIGH WATER MARK:
  35.8% qp_Cstar_mirror
  35.8% qp_Mstar_mirror
  10.9% shape_interp_mirror
  10.9% deriv_interp_mirror
  1.5% node_x0_mirror
  1.5% node_u_mirror
  1.3% node_u_dot_mirror
  1.3% node_u_ddot_mirror
  1.0% qp_weight_mirror

KOKKOS CUDA SPACE:
===================
MAX MEMORY ALLOCATED: 14056997.1 kB
ALLOCATIONS AT TIME OF HIGH WATER MARK:
  25.9% Step/Solve System/KokkosLapack::gesv[TPL_CUSOLVER,double]/getrf workspace
  25.9% St
  21.8% K
  21.8% T
  2.1% Step/Assemble System/CopyIntoSparseMatrix/Kokkos::CudaSpace::TeamScratchMemory
  2.0% gradient_matrix
```

**Comment by michaelasprague** [comment]
Ready to go; waiting for review, @deslaughter @faisal-bhuiyan
